program/tests/interface.rs

```diff
@@ -0,0 +1,830 @@
+    fn resolve_accounts(&self, account_metas: &[AccountMeta]) -> Vec<(Pubkey, Account)> {
+        let mut accounts = vec![];
+        for account_meta in account_metas {
+            let key = account_meta.pubkey;
+            let account_shared_data = if Rent::check_id(&key) {
+                self.mollusk.sysvars.keyed_account_for_rent_sysvar().1
+            } else if Clock::check_id(&key) {
+                self.mollusk.sysvars.keyed_account_for_clock_sysvar().1
+            } else if EpochSchedule::check_id(&key) {
+                self.mollusk
+                    .sysvars
+                    .keyed_account_for_epoch_schedule_sysvar()
+                    .1
+            } else if EpochRewards::check_id(&key) {
+                self.mollusk
+                    .sysvars
+                    .keyed_account_for_epoch_rewards_sysvar()
+                    .1
+            } else if StakeHistory::check_id(&key) {
+                self.mollusk
+                    .sysvars
+                    .keyed_account_for_stake_history_sysvar()
+                    .1
+            } else if let Some(account) = self.override_accounts.get(&key).cloned() {
+                account
+            } else {
+                self.base_accounts.get(&key).cloned().unwrap_or_default()
+            };
+
+            accounts.push((key, account_shared_data));
+        }
+
+        accounts
+    }
+
+    // immutable process that should succeed
+    fn process_success(&self, instruction: &Instruction) {
+        let accounts = self.resolve_accounts(&instruction.accounts);
+        self.mollusk
+            .process_and_validate_instruction(instruction, &accounts, &[Check::success()]);
+    }
+
+    // immutable process that should fail
+    fn process_fail(&self, instruction: &Instruction) {
+        let accounts = self.resolve_accounts(&instruction.accounts);
+        let result = self.mollusk.process_instruction(instruction, &accounts);
+        assert!(result.program_result.is_err());
+    }
+
+    fn reset(&mut self) {
+        self.override_accounts.clear()
+    }
+}
+
+// NOTE we skip:
+// * redelegate: will never be enabled
+// * minimum delegation: cannot fail in any nontrivial way
+// * deactivate delinquent: requires no signers, and our only failure test is missing signers
+// the first two need never be added but the third should be when we have more tests
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, Arbitrary)]
+enum StakeInterface {
+    Initialize(LockupState),
+    InitializeChecked,
+    Authorize(bool, AuthorityType, LockupState),
+    AuthorizeWithSeed(bool, AuthorityType, LockupState),
+    SetLockup(bool, LockupState, LockupState),
+    DelegateStake(LockupState),
+    Split(LockupState, bool),
+    Merge(LockupState),
+    MoveStake(LockupState, bool, bool),
+    MoveLamports(LockupState, bool, Option<bool>),
+    Withdraw(Option<bool>, LockupState, bool),
+    Deactivate(LockupState),
```

Comment:
nit: For readability, can you change the data from tuples to structs? ie

```
Authorize {
    checked: bool,
    authority_type: AuthorityType,
    lockup_state: LockupState
}
```

Reply:
will do! but to explain my thinking, `to_instruction()` is the only place where i felt these terms "really" needed to be defined, the `StakeInterface` struct itself is just the types needed for `Arbitrary` to generate all valid values (also the reason for `Option<bool>` as just "an arbitrary ternary value")